Library/Homebrew/formula.rb

```diff
@@ -476,6 +476,29 @@ def compiler_failures
     active_spec.compiler_failures
   end
 
+  # The best spec we can upgrade the {Formula} to so it can satisfy dependency.
+  def best_upgrade_spec_sym
```

Comment:
I think `HEAD` should be updated to stable when we can detect it's too outdated i.e. when stable have changed since we last installed `HEAD`. Thoughts? Can be easily modified to upgrading to the latest `HEAD` though.

> Maybe add this as a comment above this method (if I've understood correctly)

👍 I will, when we're happy with the approach.

> I also think linking doesn't matter but opt-linking does.

What do you mean?

> I think it should probably live on Dependency and be called something like def oldest_spec_for_dependency_upgrade

Initially the motivation behind placing this method into `Formula` was that we don't need to change `Dependency` in any way if we decide that we need some other logic to resolve specs for dependencies. That said, the same way we can just modify that single method inside `Dependency` and get that new behaviour. Maybe will think about a shorter name and will put it to `Dependency`.